### PR TITLE
[G2M] buttons - add button base component

### PR DESCRIPTION
**Pull request by hyx131**

### CODE REVIEW

File: src/components/button.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    }),
+  }
+
+  const classes = {
+    button: clsx(`focus:outline-none rounded-sm ${sizes.text[size]} font-body font-normal tracking-wide ${variants[variant]}`, {
+      [sizes[size]]: typeof children === 'string',
+      [sizes.squared[size]]: typeof children !== 'string',
+    }),
+    startIcon: 'pr-2.5',
+    endIcon: 'pl-2.5',
```
Comment (woozyking) [inline src/components/button.js:85]: start/end icon padding should scale with button size and content padding (px/py)

chromatic comment https://www.chromatic.com/pullrequest?appId=60d0cd6c298707003972699d&number=6&activeElementId=comment-thread-60d292b14037be00397c17aa

**Review by woozyking — CHANGES_REQUESTED** [review]

**Review by sallkall — CHANGES_REQUESTED** [review]
letter-spacing is inconsistent with the design ([chromatic thread](https://www.chromatic.com/pullrequest?appId=60d0cd6c298707003972699d&number=6&activeElementId=comment-thread-60d29a05b9a893003bcb661c))

File: src/components/button.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+import { ButtonBase } from '../base-components'
+import clsx from 'clsx'
+
+
+const buttonConfigs = () => ({
+  colors: {
```
Comment (kc-leung) [inline src/components/button.js:9]: I think for color it is better to be as dynamic as possible. I can't imagine having a whole palette inside of button.js in buttonConfigs

**Comment by kc-leung** [comment]
we should start creating component's own css file or at least a public one. In the case I need to add a new styling and I do it in tailwind.config.js it will cause a lot of conflict code which is gonna be really hard to review and fix since it is a very large file

File: src/components/button.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      [sizes[size]]: typeof children === 'string',
+      [sizes.squared[size]]: typeof children !== 'string',
+    }),
+    startIcon: sizes.iconPadding.startIcon[size],
+    endIcon: sizes.iconPadding.endIcon[size],
+  }
+
+  return (
+    <ButtonBase classes={classes} disabled={disabled} {...rest}>
+      {(typeof children === 'string' && size === 'sm') ? children.toUpperCase() : children}
```
Comment (woozyking) [inline src/components/button.js:102]: not sure about performance/memory consumption between the CSS text transformation vs. JS string transformation.

however, take into consideration of misuse of this component, where `children` can be anything, _not_ always a string (and naturally so as a good children-wrapper component), such as the following case shouldn't be constrained:

```jsx
<Button>
    <span>lol</span>
</Button>
```

in such case, CSS text transformation becomes a bit more appealing as a more robust solution:

```jsx
const ButtonBase = (...) => {
  <button className={{..., size determination logic to apply tailwind's .uppercase class or not }} ...>
    ...
  </button>
}
```

may not be the best place, perhaps simpler to wrap the `children` inside the `Button` component with a transient layer of `<span>` that allows you to apply this class:

```jsx
const Button = (...) => {
  ...
  return (
    <ButtonBase ...>
      <span className={size === 'sm' ? 'uppercase' : ''}>
        {children}
      </span>
    </ButtonBase>
  )
} 
```

or wrap the condition outside of `<span>` to render `children` directly if the `size` isn't `sm`

**Comment by woozyking** [comment]
> we should start creating component's own css file or at least a public one. In the case I need to add a new styling and I do it in tailwind.config.js it will cause a lot of conflict code which is gonna be really hard to review and fix since it is a very large file

this is worth considering

we can also mix-in css-in-js where it's due (especially to work with dynamic logic on custom styles), some lightweight alternatives such as [goober](https://goober.js.org/) can be considered (much lighter than `styled-component` and `emotion`, for example).

or we can even look into build-time css generator solution such as [treat](https://seek-oss.github.io/treat/) which doesn't bring the usual performance trade-off that css-in-js comes. however, this invites another monster that may overcomplex the project development (webpack, or any applicable equivalent bundler)

File: src/components/button.js (hunk 3 above)
Comment (hyx131) [reply src/components/button.js:102]: ohh oki, will change to css text-transformation

File: src/components/button.js (hunk 2 above)
Comment (hyx131) [reply src/components/button.js:9]: sorry for this massive button configs object, i'll try to see if i could reduce it a bit
but this isn't the color palette, the color palette is in `tailwind.config.js` that looks something like this:
```
      'secondary-50': '#fafafa',
      'secondary-100': '#f5f5f5',
      'secondary-200': '#eeeeee',
      'secondary-300': '#e0e0e0',
      'secondary-400': '#bdbdbd',
      'secondary-500': '#9e9e9e',
      'secondary-600': '#757575',
      'secondary-700': '#616161',
      'secondary-800': '#424242',
      'secondary-900': '#212121',

      success: '#00D308',
      'success-hover': '#00b306',
```
this field of the config is for the `color` prop, which includes color types supported from our design system `default/normal/warning/error`, as well as the disabled/hovering effects of each variant (`filled/outlined/shaded/borderless`) for each color type
if none of these color types match what the user of these buttons is looking for, then they could (1) import the ButtonBase component & apply their own styles, or (2) change just the color via `className` (i'll open up options to override colors/borders/shadings/etc)

**Comment by hyx131** [comment]
> > we should start creating component's own css file or at least a public one. In the case I need to add a new styling and I do it in tailwind.config.js it will cause a lot of conflict code which is gonna be really hard to review and fix since it is a very large file
> 
> this is worth considering, but definitely prioritize on using tailwind classes whenever appropriate
> 
> we can also mix-in css-in-js where it's due (especially to work with dynamic logic on custom styles), some lightweight alternatives such as [goober](https://goober.js.org/) can be considered (much lighter than `styled-component` and `emotion`, for example).
> 
> or we can even look into build-time css generator solution such as [treat](https://seek-oss.github.io/treat/) which doesn't bring the usual performance trade-off that css-in-js comes. however, this invites another monster that may overcomplicate the project development (webpack, or any applicable equivalent bundler)

for sure, the `tailwind.config.js` includes style themes (e.g. for colors, spacing, fonts, shadings, etc), the idea is to remove unnecessary configs from `tailwind.config.js` once our own theming system is finalized. With theming in place, plus the tailwind classes, would assume there won't be a lot of cases where we'd need a separate css file (suggest to add in this separate file only when needed)

**Comment by woozyking** [comment]
> would assume there won't be a lot of cases where we'd need a separate css file (suggest to add in this separate file only when needed)

yep, fits the tiered system proposed:
1. base stylings for primitives (tailwind covers this)
2. css classes (tailwind + our theming configs cover this)
3. ad-hoc styling (static or css-in-js)

**Review by woozyking — APPROVED** [review]

**Review by sallkall — APPROVED** [review]

**Pull request merged**